openlibrary/core/bookshelves.py

```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
+                    logger.warning(
+                        "Error: No work found for edition %s" % edition["key"]
+                    )
+        return linked_docs
+
+    @classmethod
+    def add_storage_items_for_redirects(
+        cls, reading_log_keys, solr_docs: list[web.Storage]
+    ) -> list[web.storage]:
+        """
+        Use reading_log_keys to fill in missing redirected items in the
+        the solr_docs query results.
+
+        Solr won't return matches for work keys that have been redirected. Because
+        we use Solr to build the lists of storage items that ultimately gets passed
+        to the templates, redirected items returned from the reading log DB will
+        'disappear' when not returned by Solr. This remedies that by filling in
+        dummy works, albeit with the correct work_id.
+        """
+
+        from openlibrary.plugins.worksearch.code import run_solr_query
+        from openlibrary.plugins.worksearch.schemes.works import WorkSearchScheme
+
+        fetched_keys = {doc["key"] for doc in solr_docs}
+        missing_keys = {work for (work, _) in reading_log_keys} - fetched_keys
+
+        """
+        Provides a proper 1-to-1 connection between work keys and edition keys; needed, in order to fill in the appropriate 'logged_edition' data
+        for the correctly pulled work later on, as well as to correctly update the post-redirect version back to the pre_redirect key.
+        Without this step, processes may error due to the 'post-redirect key' not actually existing within a user's reading log.
+        """
+        work_to_edition_keys = {
+            work: edition for (work, edition) in reading_log_keys if edition
+        }
+        edition_to_work_keys = {
+            edition.split("/")[2]: work
+            for (work, edition) in reading_log_keys
+            if edition
+        }
+
+        # Here, we add in dummied works for situations in which there is no edition key present, yet the work key accesses a redirect.
+        # Ideally, this will be rare, as there's no way to access the relevant information through Solr.
+        for key in missing_keys.copy():
+            if not work_to_edition_keys.get(key):
+                missing_keys.remove(key)
+                solr_docs.append(web.storage({"key": key}))
+
+        keys_to_fetch = [
+            work_to_edition_keys[key].split("/")[2] for key in missing_keys
+        ]
+        if fq := f'edition_key:{" OR ".join(keys_to_fetch)}':
+            solr_resp = run_solr_query(
+                scheme=WorkSearchScheme(),
+                param={'q': '*:*'},
+                fields=list(
+                    WorkSearchScheme.default_fetched_fields
+                    | {'subject', 'person', 'place', 'time', 'edition_key'}
+                ),
+                facet=False,
+                extra_params=[("fq", fq)],
+            ).docs
```

Comment:
Drop `docs` here so `solr_resp` below actually refers to a `SolrResponse` object

```suggestion
            )
```

Reply:
Fixed, thanks for the advice!